File: src/SaruchSDK/Interfaces/FolderClient.vb
Imports Newtonsoft.Json
Imports SaruchSDK.JSON
Imports SaruchSDK.utilitiez

Public Class FolderClient
    Implements IFolder

    Private Property FolderID As String

    Sub New(FolderID As String)
        Me.FolderID = FolderID
    End Sub
    Sub New()
    End Sub

#Region "List"
    Public Async Function GET_List(Optional OffSet As Integer = 1, Optional Sort As SortEnum = SortEnum.name, Optional OrderBy As OrderByEnum = OrderByEnum.asc) As Task(Of JSON_List) Implements IFolder.List
        Using localHttpClient As New HttpClient(New HCHandler)
            Dim RequestUri = New pUri("video-manager", New Dictionary(Of String, String) From {{"per_page", 50}, {"folder_id", FolderID}, {"page", OffSet}, {"sort", JsonConvert.SerializeObject(New With {.where = Sort.ToString, .by = OrderBy.ToString})}})
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(RequestUri).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Return JsonConvert.DeserializeObject(Of JSON_List)(result, JSONhandler)
                Else
                    result.ShowError()
                End If
            End Using
        End Using
    End Function
#End Region

#Region "Metadata"
    Public Async Function _Metadata() As Task(Of JSON_FolderMetadata) Implements IFolder.Metadata
        FolderID.IsRequired()
        Using localHttpClient As New HttpClient(New HCHandler)
            Dim RequestUri = New pUri(String.Format("folders/{0}", FolderID))
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(RequestUri).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Return JsonConvert.DeserializeObject(Of JSON_GetFolderMetadata)(result, JSONhandler).folder
                Else
                    result.ShowError()
                End If
            End Using
        End Using
    End Function
#End Region

#Region "CreateNewFolder"
    Public Async Function _CreateNewFolder(FolderName As String) As Task(Of JSON_FolderMetadata) Implements IFolder.Create
        Dim parameters = New Dictionary(Of String, String) From {{"parent_id", FolderID}, {"name", FolderName}}

        Using localHttpClient As New HttpClient(New HCHandler)
            Dim HtpReqMessage As New Net.Http.HttpRequestMessage()
            HtpReqMessage.RequestUri = New pUri("folders")
            HtpReqMessage.Method = Net.Http.HttpMethod.Post
            HtpReqMessage.Content = parameters.EncodedContent
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.SendAsync(HtpReqMessage, Net.Http.HttpCompletionOption.ResponseContentRead).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Return JsonConvert.DeserializeObject(Of JSON_FolderMetadata)(result, JSONhandler)
                Else
                    result.ShowError()
                End If
            End Using
        End Using
    End Function
#End Region

#Region "RenameFolder"
    Public Async Function _RenameFolder(NewName As String) As Task(Of Boolean) Implements IFolder.Rename
        FolderID.IsRequired()
        Dim parameters = New Dictionary(Of String, String) From {{"name", NewName}}

        Using localHttpClient As New HttpClient(New HCHandler)
            Dim HtpReqMessage As New Net.Http.HttpRequestMessage()
            HtpReqMessage.RequestUri = New pUri(String.Format("folders/{0}", FolderID))
            HtpReqMessage.Method = Net.Http.HttpMethod.Put
            HtpReqMessage.Content = parameters.EncodedContent
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.SendAsync(HtpReqMessage, Net.Http.HttpCompletionOption.ResponseContentRead).ConfigureAwait(False)

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Return True
                Else
                    Dim result As String = Await response.Content.ReadAsStringAsync()
                    result.ShowError()
                End If
            End Using
        End Using
    End Function
#End Region

#Region "MoveFolder"
    Public Async Function _MoveFolder(DestinationFolderID As String) As Task(Of Boolean) Implements IFolder.Move
        FolderID.IsRequired()
        Dim parameters = New Dictionary(Of String, String) From {{"folder_id", DestinationFolderID}}

        Using localHttpClient As New HttpClient(New HCHandler)
            Dim HtpReqMessage As New Net.Http.HttpRequestMessage()
            HtpReqMessage.RequestUri = New pUri(String.Format("folders/{0}", FolderID))
            HtpReqMessage.Method = Net.Http.HttpMethod.Put
            HtpReqMessage.Content = parameters.EncodedContent
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.SendAsync(HtpReqMessage, Net.Http.HttpCompletionOption.ResponseContentRead).ConfigureAwait(False)

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Return True
                Else
                    Dim result As String = Await response.Content.ReadAsStringAsync()
                    result.ShowError()
                End If
            End Using
        End Using
    End Function
#End Region

#Region "DeleteFolder"
    Public Async Function _DeleteFolder() As Task(Of Boolean) Implements IFolder.Delete
        FolderID.IsRequired()
        Using localHttpClient As New HttpClient(New HCHandler)
            Dim RequestUri = New pUri(String.Format("folders/{0}", FolderID))
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.DeleteAsync(RequestUri).ConfigureAwait(False)

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Return True
                Else
                    Dim result As String = Await response.Content.ReadAsStringAsync()
                    result.ShowError()
                End If
            End Using
        End Using
    End Function
#End Region

#Region "UploadFile"

#Region "GetUploadUrl"
    Private Async Function _GetUploadUrl(FileName As String, FileSize As String) As Task(Of JSON_GetUploadUrl)
        If IO.Path.GetExtension(FileName).ToLower = ".wmv" Then Throw New System.Exception("Mime type not allowed")

        Dim parameters = New Dictionary(Of String, String)
        parameters.Add("name", FileName)
        parameters.Add("size", FileSize)
        parameters.Add("mime_type", "video/mp4") 'Web.MimeMapping.GetMimeMapping(FileName))
        parameters.Add("folder_id", FolderID)

        Using localHttpClient As New HttpClient(New HCHandler)
            Dim HtpReqMessage As New Net.Http.HttpRequestMessage()
            HtpReqMessage.RequestUri = New pUri("videos")
            HtpReqMessage.Method = Net.Http.HttpMethod.Post
            HtpReqMessage.Content = parameters.EncodedContent
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.SendAsync(HtpReqMessage, Net.Http.HttpCompletionOption.ResponseContentRead).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Return JsonConvert.DeserializeObject(Of JSON_GetUploadUrl)(result, JSONhandler)
                Else
                    result.ShowError()
                End If
            End Using
        End Using
    End Function
#End Region

    Public Async Function Get_UploadLocal(FileToUpload As Object, UploadType As UploadTypes, FileName As String, Optional ReportCls As IProgress(Of ReportStatus) = Nothing, Optional token As Threading.CancellationToken = Nothing) As Task(Of JSON_Upload) Implements IFolder.UploadLocalVideo
        Dim filesize As Integer = 0
        If ReportCls Is Nothing Then ReportCls = New Progress(Of ReportStatus)
        ReportCls.Report(New ReportStatus With {.Finished = False, .TextStatus = "Initializing..."})
        Try
            Dim progressHandler As New Net.Http.Handlers.ProgressMessageHandler(New HCHandler)
            AddHandler progressHandler.HttpSendProgress, (Function(sender, e)
                                                              ReportCls.Report(New ReportStatus With {.ProgressPercentage = e.ProgressPercentage, .BytesTransferred = e.BytesTransferred, .TotalBytes = If(e.TotalBytes Is Nothing, 0, e.TotalBytes), .TextStatus = "Uploading..."})
                                                          End Function)
            Dim localHttpClient As New HttpClient(progressHandler)
            Dim HtpReqMessage As New Net.Http.HttpRequestMessage()
            HtpReqMessage.Method = Net.Http.HttpMethod.Post
            ''''''''''''''''''''''''''''''''''
            Dim MultipartsformData = New Net.Http.MultipartFormDataContent()
            Dim streamContent As Net.Http.HttpContent
            Select Case UploadType
                Case UploadTypes.FilePath
                    streamContent = New Net.Http.StreamContent(New IO.FileStream(FileToUpload, IO.FileMode.Open, IO.FileAccess.Read))
                    filesize = New System.IO.FileInfo(FileToUpload).Length
                Case UploadTypes.Stream
                    Dim strm = CType(FileToUpload, IO.Stream)
                    streamContent = New Net.Http.StreamContent(strm)
                    filesize = strm.Length
                Case UploadTypes.BytesArry
                    Dim byt = CType(FileToUpload, Byte())
                    streamContent = New Net.Http.StreamContent(New IO.MemoryStream(byt))
                    filesize = byt.Length
            End Select
            streamContent.Headers.Clear()
            streamContent.Headers.ContentDisposition = New System.Net.Http.Headers.ContentDispositionHeaderValue("form-data") With {.Name = """video""", .FileName = """" + FileName + """"}
            ''''''''''''''''''''''''''
            Dim uploadUrl = Await _GetUploadUrl(FileName, filesize)
            MultipartsformData.Add(New Net.Http.StringContent(uploadUrl.video.id), """video_id""")
            MultipartsformData.Add(New Net.Http.StringContent("resumeable"), """upload_type""")
            MultipartsformData.Add(streamContent)
            HtpReqMessage.Content = MultipartsformData

            HtpReqMessage.RequestUri = New Uri(uploadUrl.server)
            '''''''''''''''''will write the whole content to H.D WHEN download completed'''''''''''''''''''''''''''''
            Using ResPonse As Net.Http.HttpResponseMessage = Await localHttpClient.SendAsync(HtpReqMessage, Net.Http.HttpCompletionOption.ResponseHeadersRead, token).ConfigureAwait(False)
                Dim result As String = Await ResPonse.Content.ReadAsStringAsync()

                token.ThrowIfCancellationRequested()
                If ResPonse.StatusCode = Net.HttpStatusCode.OK Then
                    ReportCls.Report(New ReportStatus With {.Finished = True, .TextStatus = String.Format("[{0}] Uploaded successfully", FileName)})
                    Return JsonConvert.DeserializeObject(Of JSON_Upload)(result, JSONhandler)
                Else
                    ShowError(result)
                    ReportCls.Report(New ReportStatus With {.Finished = True, .TextStatus = String.Format("The request returned with HTTP status code {0}", ResPonse.ReasonPhrase)})
                End If
            End Using
        Catch ex As Exception
            ReportCls.Report(New ReportStatus With {.Finished = True})
            If ex.Message.ToString.ToLower.Contains("a task was canceled") Then
                ReportCls.Report(New ReportStatus With {.TextStatus = ex.Message})
            Else
                Throw ExceptionCls.CreateException(ex.Message, Nothing)
            End If
        End Try
    End Function
#End Region

#Region "RemoteUpload"
    Public Async Function GET_RemoteUpload(FileUrls As List(Of String)) As Task(Of JSON_RemoteUpload) Implements IFolder.UploadRemoteVideo
        Using localHttpClient As New HttpClient(New HCHandler)
            Dim HtpReqMessage As New Net.Http.HttpRequestMessage()
            HtpReqMessage.RequestUri = New pUri("remote-uploads")
            HtpReqMessage.Method = Net.Http.HttpMethod.Post
            Dim JSONobj = New With {.urls = String.Join("\", FileUrls)}
            Dim streamContent As Net.Http.HttpContent = New Net.Http.StringContent(Newtonsoft.Json.JsonConvert.SerializeObject(JSONobj), System.Text.Encoding.UTF8, "application/json")
            HtpReqMessage.Content = streamContent
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.SendAsync(HtpReqMessage, Net.Http.HttpCompletionOption.ResponseContentRead).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Return JsonConvert.DeserializeObject(Of JSON_RemoteUpload)(result, JSONhandler)
                Else
                    ShowError(result)
                End If
            End Using
        End Using
    End Function
#End Region


End Class

File: SaruchSDK/SaruchSDK/Interfaces/FolderClient.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using SaruchSDK.JSON;
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using static SaruchSDK.Basic;
using static SaruchSDK.Utilitiez;

namespace SaruchSDK
{
    public class FolderClient : IFolder
    {
        private string FolderID { get; set; }
        public FolderClient(string FolderID) => this.FolderID = FolderID;


        public async Task<JSON_List> List(int OffSet = 1, SortEnum Sort = SortEnum.name, OrderByEnum OrderBy = OrderByEnum.asc)
        {
            HttpResponseMessage response = await RequestAsync(HttpMethod.Get, Build("video-manager", new Dictionary<string, string>() { { "per_page", "50" }, { "folder_id", FolderID }, { "page", OffSet.ToString() }, { "sort", JsonConvert.SerializeObject(new { where = Sort.ToString(), by = OrderBy.ToString() }) } }));
            string result = await response.Content.ReadAsStringAsync();
            return response.Success() ? JsonConvert.DeserializeObject<JSON_List>(result, JSONhandler) : throw ShowError(result);
        }

        public async Task<JSON_FolderMetadata> Metadata()
        {
            HttpResponseMessage response = await RequestAsync(HttpMethod.Get, new pUri($"folders/{FolderID}").ToString());
            string result = await response.Content.ReadAsStringAsync();
            return response.Success() ? JsonConvert.DeserializeObject<JSON_FolderMetadata>(JObject.Parse(result).SelectToken("folder").ToString(), JSONhandler) : throw ShowError(result);
        }

        public async Task<JSON_FolderMetadata> Create(string FolderName)
        {
            var parameters = new Dictionary<string, string>() { { "parent_id", FolderID }, { "name", FolderName } };

            HttpResponseMessage response = await RequestAsync(HttpMethod.Post, new pUri("folders").ToString(), parameters.EncodedContent());
            string result = await response.Content.ReadAsStringAsync();
            return response.Success() ? JsonConvert.DeserializeObject<JSON_FolderMetadata>(result, JSONhandler) : throw ShowError(result);
        }

        public async Task<bool> Rename(string NewName)
        {
            var parameters = new Dictionary<string, string>() { { "name", NewName } };

            HttpResponseMessage response = await RequestAsync(HttpMethod.Put, new pUri($"folders/{FolderID}").ToString(), parameters.EncodedContent());
            string result = await response.Content.ReadAsStringAsync();
            return response.Success() ? true : throw ShowError(result);
        }

        public async Task<bool> Move(string DestinationFolderID)
        {
            var parameters = new Dictionary<string, string>() { { "folder_id", DestinationFolderID } };
            HttpResponseMessage response = await RequestAsync(HttpMethod.Put, new pUri($"folders/{FolderID}").ToString(), parameters.EncodedContent());
            string result = await response.Content.ReadAsStringAsync();
            return response.Success() ? true : throw ShowError(result);
        }

        public async Task<bool> Delete()
        {
            HttpResponseMessage response = await RequestAsync(HttpMethod.Delete, new pUri($"folders/{FolderID}").ToString());
            string result = await response.Content.ReadAsStringAsync();
            return response.Success() ? true : throw ShowError(result);
        }

        private async Task<JSON_GetUploadUrl> GetUploadUrl(string FileName, string FileSize)
        {
            if (Path.GetExtension(FileName).ToLower() == ".wmv"){throw new SaruchException ("Mime type not allowed",405);}

            var parameters = new Dictionary<string, string>
            {
                { "name", FileName },
                { "size", FileSize },
                { "mime_type", "video/mp4" }, // Web.MimeMapping.GetMimeMapping(FileName))
                { "folder_id", FolderID }
            };

            HttpResponseMessage response = await RequestAsync(HttpMethod.Post, new pUri("videos").ToString(), parameters.EncodedContent());
            string result = await response.Content.ReadAsStringAsync();
            return response.Success() ? JsonConvert.DeserializeObject<JSON_GetUploadUrl>(result, JSONhandler) : throw ShowError(result);
        }

        public async Task<JSON_Upload> Upload(object FileToUpload, UploadTypes UploadType, string FileName, IProgress<ReportStatus> ReportCls = null, CancellationToken token = default)
        {
            long filesize = 0;
            ReportCls = ReportCls ?? new Progress<ReportStatus>();
            ReportCls.Report(new ReportStatus() { Finished = false, TextStatus = "Initializing..." });
            try
            {
                System.Net.Http.Handlers.ProgressMessageHandler progressHandler = new System.Net.Http.Handlers.ProgressMessageHandler(new HCHandler());
                progressHandler.HttpSendProgress += (sender, e) => { ReportCls.Report(new ReportStatus() { ProgressPercentage = e.ProgressPercentage, BytesTransferred = e.BytesTransferred, TotalBytes = e.TotalBytes ?? 0, TextStatus = "Uploading..." }); };
                using (HtpClient localHttpClient = new HtpClient(progressHandler))
                {
                    HttpRequestMessage HtpReqMessage = new HttpRequestMessage();
                    HtpReqMessage.Method = HttpMethod.Post;
                    var MultipartsformData = new MultipartFormDataContent();
                    HttpContent streamContent = null;
                    switch (UploadType)
                    {
                        case UploadTypes.FilePath:
                            streamContent = new StreamContent(new FileStream(FileToUpload.ToString(), FileMode.Open, FileAccess.Read));
                            filesize = new FileInfo(FileToUpload.ToString()).Length;
                            break;
                        case UploadTypes.Stream:
                            var strm = (Stream)FileToUpload;
                            streamContent = new StreamContent(strm);
                            filesize = strm.Length;
                            break;
                        case UploadTypes.BytesArry:
                            byte[] byt = (byte[])FileToUpload;
                            streamContent = new StreamContent(new MemoryStream(byt));
                            filesize = byt.Length;
                            break;
                    }
                    streamContent.Headers.Clear();
                    streamContent.Headers.ContentDisposition = new System.Net.Http.Headers.ContentDispositionHeaderValue("form-data") { Name = "\"video\"", FileName = "\"" + FileName + "\"" };
                    // '''''''''''''''''''''''''
                    var uploadUrl = await GetUploadUrl(FileName, filesize.ToString());
                    MultipartsformData.Add(new StringContent(uploadUrl.video.id), "\"video_id\"");
                    MultipartsformData.Add(new StringContent("resumeable"), "\"upload_type\"");
                    MultipartsformData.Add(streamContent);
                    HtpReqMessage.Content = MultipartsformData;
                    HtpReqMessage.RequestUri = new Uri(uploadUrl.server);
                    // ''''''''''''''''will write the whole content to H.D WHEN download completed'''''''''''''''''''''''''''''
                    using (HttpResponseMessage ResPonse = await localHttpClient.SendAsync(HtpReqMessage, HttpCompletionOption.ResponseHeadersRead, token).ConfigureAwait(false))
                    {
                        string result = await ResPonse.Content.ReadAsStringAsync();

                        token.ThrowIfCancellationRequested();
                        if (ResPonse.StatusCode == System.Net.HttpStatusCode.OK)
                        {
                            ReportCls.Report(new ReportStatus() { Finished = true, TextStatus = $"[{FileName}] Uploaded successfully" });
                            return JsonConvert.DeserializeObject<JSON_Upload>(result, JSONhandler);
                        }
                        else
                        {
                            ReportCls.Report(new ReportStatus() { Finished = true, TextStatus = $"The request returned with HTTP status code {ResPonse.ReasonPhrase}" });
                            throw ShowError(result);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                ReportCls.Report(new ReportStatus() { Finished = true });
                if (ex.Message.ToString().ToLower().Contains("a task was canceled"))
                {
                    ReportCls.Report(new ReportStatus() { TextStatus = ex.Message });
                }
                else
                {
                    throw new SaruchException (ex.Message, 1001);
                }
                return null;
            }
        }

        public async Task<JSON_RemoteUpload> UploadRemotly(List<string> FileUrls)
        {
            var JSONobj = new { urls = string.Join(@"\", FileUrls) };
            HttpContent streamContent = new StringContent(Newtonsoft.Json.JsonConvert.SerializeObject(JSONobj), System.Text.Encoding.UTF8, "application/json");
            HttpResponseMessage response = await RequestAsync(HttpMethod.Post, new pUri("remote-uploads").ToString(), streamContent);
            string result = await response.Content.ReadAsStringAsync();
            return response.Success() ? JsonConvert.DeserializeObject<JSON_RemoteUpload>(result, JSONhandler) : throw ShowError(result);
        }

    }
}
